### PR TITLE
Update std iterators

### PR DESCRIPTION
**Pull request by miscco**
This updates the non exposed standard iterators to the new c++20 state

### CODE REVIEW

File: libcudacxx/include/cuda/std/detail/libcxx/include/__iterator/back_insert_iterator.h [hunk 1]
```diff
@@ -35,31 +35,39 @@ _LIBCUDACXX_SUPPRESS_DEPRECATED_PUSH
...
-                      void,
-                      void,
-                      void,
-                      void>
+    : public iterator<output_iterator_tag, void, void, void, void>
 #endif
 {
 _LIBCUDACXX_SUPPRESS_DEPRECATED_POP
 protected:
     _Container* container;
```
Comment (gevtushenko) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__iterator/back_insert_iterator.h:43]: question: in the standard, this seems to be `_Container* container = nullptr;`. Having it initialized might help diagnoze issues. Is there anything preventing us from initializing pointer?

I'm also unable to find default constructor:
```cpp
constexpr back_insert_iterator() noexcept = default;
```
Should there be one? Same questions apply to remaining iterators.

**Review by gevtushenko — APPROVED** [review]

File: libcudacxx/include/cuda/std/detail/libcxx/include/__iterator/back_insert_iterator.h (hunk 1 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__iterator/back_insert_iterator.h:43]: So we intentionally do not provide a default constructor exactly for the reason that we always want `container` to be properly initialized: https://en.cppreference.com/w/cpp/iterator/back_insert_iterator/back_insert_iterator

Given that this is essentially dead code from our libcu++ fork I would rather keep it as is and close to upstream

Comment (gevtushenko) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__iterator/back_insert_iterator.h:43]: Oh, it seems that I was looking at older draft, thank you for the reference!

File: libcudacxx/include/cuda/std/detail/libcxx/include/__iterator/back_insert_iterator.h [hunk 2]
```diff
@@ -35,31 +35,39 @@ _LIBCUDACXX_SUPPRESS_DEPRECATED_PUSH
...
 public:
+    typedef output_iterator_tag iterator_category;
```
[the remark this answers is not in the document]
Comment (miscco) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__iterator/back_insert_iterator.h:45]: This header is only used for the libcxx tests and not exposed in libcu++. So I want to keep it as close as possible to source

**Pull request merged**